### PR TITLE
Remove derived metrics and support a callback for the entire metric.

### PR DESCRIPTION
**Pull request by bogdandrutu**
Based on the yesterday feedback about metrics, trying to simplify the API:
* Removed Derived metrics, and add support to set a callback for a metric that is called before the metric is exported:
  * This saved a lot of calls to get memory see in the contrib area metric.
  * This allows to add timeseries that we don't know in advance (like all the pool names or gc names).
* It would be good to support a callback at Registry (collection) level to allow updating multiple metrics from that collection - see the memory example where I added a type mostly to avoid 3 separate callbacks to be executed (there are other advantages in this case so we may keep the type even after we add that callback).

### CODE REVIEW

File: contrib/src/main/java/openconsensus/contrib/metrics/runtime/GarbageCollector.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
+package openconsensus.contrib.metrics.runtime;
```
Comment (songy23) [inline contrib/src/main/java/openconsensus/contrib/metrics/runtime/GarbageCollector.java]: Nit: license header. Also probably better to split JVM metrics to another PR.

Comment (bogdandrutu) [reply contrib/src/main/java/openconsensus/contrib/metrics/runtime/GarbageCollector.java]: JVM metrics are an example of usage to prove the new design choice. Also the contrib package in this repo is mostly to prove what we can build on top of the API. We need tests and better review before we release that.

**Comment by rghetia** [comment]
LGTM as well.

File: api/src/main/java/openconsensus/metrics/Metric.java [hunk 2]
```diff
@@ -21,7 +21,44 @@
...
+  /**
+   * Creates a {@code TimeSeries} and returns a {@code TimeSeries} if the specified {@code
+   * labelValues} is not already associated with this gauge, else returns an existing {@code
+   * TimeSeries}.
+   *
+   * <p>It is recommended to keep a reference to the TimeSeries instead of always calling this
+   * method for every operations.
+   *
+   * @param labelValues the list of label values. The number of label values must be the same to
+   *     that of the label keys passed to {@link DoubleGauge.Builder#setLabelKeys(List)}.
```
Comment (c24t) [inline api/src/main/java/openconsensus/metrics/Metric.java]: Some copypasta here

File: api/src/main/java/openconsensus/metrics/Metric.java [hunk 3]
```diff
@@ -21,7 +21,44 @@
 import java.util.Map;
 
 /** Base interface for all metrics defined in this package. */
-public interface Metric {
+public interface Metric<T> {
+  /**
+   * Creates a {@code TimeSeries} and returns a {@code TimeSeries} if the specified {@code
+   * labelValues} is not already associated with this gauge, else returns an existing {@code
```
Comment (songy23) [inline api/src/main/java/openconsensus/metrics/Metric.java]: s/gauge/metric/g

File: api/src/main/java/openconsensus/metrics/DoubleGauge.java [hunk 4]
```diff
@@ -75,33 +76,12 @@
  * @since 0.1.0
  */
 @ThreadSafe
-public interface DoubleGauge extends Metric {
+public interface DoubleGauge extends Metric<TimeSeries> {
```
Comment (c24t) [inline api/src/main/java/openconsensus/metrics/DoubleGauge.java]: Talking to @bogdandrutu, the reason we have separate classes for double and long gauges/cumulatives instead of a single generic class is to avoid `Int` and `Double` types and the headache of java type boxing. Is there a better way around this?

File: api/src/main/java/openconsensus/metrics/DoubleCumulative.java [hunk 5]
```diff
@@ -75,33 +75,12 @@
  * @since 0.1.0
  */
 @ThreadSafe
-public interface DoubleCumulative extends Metric {
+public interface DoubleCumulative extends Metric<DoubleCumulative.TimeSeries> {
```
Comment (c24t) [inline api/src/main/java/openconsensus/metrics/DoubleCumulative.java]: Last chance to call these `Accumulator`s instead of `Cumulative`s to avoid nouning the adjective. :D

File: api/src/main/java/openconsensus/metrics/LongCumulative.java [hunk 6]
```diff
@@ -75,33 +76,12 @@
  * @since 0.1.0
  */
 @ThreadSafe
-public interface LongCumulative extends Metric {
+public interface LongCumulative extends Metric<TimeSeries> {
```
Comment (c24t) [inline api/src/main/java/openconsensus/metrics/LongCumulative.java]: Should this be `LongCumulative.TimeSeries`?

File: api/src/main/java/openconsensus/metrics/LongCumulative.java [hunk 7]
```diff
@@ -118,6 +98,17 @@
...
+    /**
+     * Sets the given value. The value must be larger than the current recorded value.
+     *
+     * <p>In general should be used in combination with {@link #setCallback(Runnable)} where the
+     * recorded value is guaranteed to be monotonically increasing.
+     *
+     * @param val the new value.
+     * @since 0.1.0
+     */
+    void set(long val);
```
Comment (c24t) [inline api/src/main/java/openconsensus/metrics/LongCumulative.java]: Can you add `set` to `TimeSeries` and remove the special cumulative `TimeSeries` interfaces?

File: api/src/main/java/openconsensus/metrics/LongCumulative.java [hunk 8]
```diff
@@ -118,6 +98,17 @@
      * @since 0.1.0
      */
     void add(long delta);
+
+    /**
+     * Sets the given value. The value must be larger than the current recorded value.
```
[the remark this answers is not in the document]
Comment (bogdandrutu) [reply api/src/main/java/openconsensus/metrics/LongCumulative.java]: Done.

File: api/src/main/java/openconsensus/metrics/DoubleCumulative.java [hunk 9]
```diff
@@ -118,6 +97,17 @@
      * @since 0.1.0
      */
     void add(double delta);
+
+    /**
+     * Sets the given value. The value must be larger than the current recorded value.
```
[the remark this answers is not in the document]
Comment (bogdandrutu) [reply api/src/main/java/openconsensus/metrics/DoubleCumulative.java]: Done.

File: api/src/main/java/openconsensus/metrics/Metric.java (hunk 3 above)
Comment (bogdandrutu) [reply api/src/main/java/openconsensus/metrics/Metric.java]: done.

File: api/src/main/java/openconsensus/metrics/Metric.java [hunk 10]
```diff
@@ -21,7 +21,44 @@
...
+   * TimeSeries}.
+   *
+   * <p>It is recommended to keep a reference to the TimeSeries instead of always calling this
+   * method for every operations.
```
[the remark this answers is not in the document]
Comment (bogdandrutu) [reply api/src/main/java/openconsensus/metrics/Metric.java]: Done.

File: api/src/main/java/openconsensus/metrics/Metric.java (hunk 2 above)
Comment (bogdandrutu) [reply api/src/main/java/openconsensus/metrics/Metric.java]: Done.

File: api/src/main/java/openconsensus/metrics/LongCumulative.java (hunk 6 above)
Comment (bogdandrutu) [reply api/src/main/java/openconsensus/metrics/LongCumulative.java]: I had an import in the file for LongCumulative.TimeSeries, but changed it to be consistent.

File: api/src/main/java/openconsensus/metrics/LongCumulative.java (hunk 7 above)
Comment (bogdandrutu) [reply api/src/main/java/openconsensus/metrics/LongCumulative.java]: not really because that forces boxing (same reason we have Double/Long Metrics).

File: api/src/main/java/openconsensus/metrics/DoubleCumulative.java (hunk 5 above)
Comment (bogdandrutu) [reply api/src/main/java/openconsensus/metrics/DoubleCumulative.java]: Some systems call this counter.

File: api/src/main/java/openconsensus/metrics/Metric.java [hunk 11]
```diff
@@ -21,7 +21,44 @@
 import java.util.Map;
 
 /** Base interface for all metrics defined in this package. */
-public interface Metric {
+public interface Metric<T> {
```
[the remark this answers is not in the document]
Comment (bogdandrutu) [reply api/src/main/java/openconsensus/metrics/Metric.java]: Done.

File: api/src/main/java/openconsensus/metrics/DoubleCumulative.java (hunk 5 above)
Comment (SergeyKanzhelev) [reply api/src/main/java/openconsensus/metrics/DoubleCumulative.java]: Counters is definitely easier to explain
